signer/xpi/x509.go

```diff
@@ -7,12 +7,63 @@ import (
 	"crypto/rsa"
 	"crypto/x509"
 	"crypto/x509/pkix"
+	"log"
 	"math/big"
 	"time"
 
 	"github.com/pkg/errors"
 )
 
+// every minute, add an rsa key to the cache. This will block if
+// the cache channel is already full, which is what we want anyway
+func (s *PKCS7Signer) populateRsa2048Cache() {
+	for {
+		key, err := rsa.GenerateKey(rand.Reader, 2048)
+		if err != nil {
+			log.Fatalf("xpi.populateRsa2048Cache: %v", err)
+		}
+		s.rsa2048Cache <- key
+		time.Sleep(time.Minute)
+	}
+}
+
+func (s *PKCS7Signer) populateRsa4096Cache() {
+	for {
+		key, err := rsa.GenerateKey(rand.Reader, 4096)
+		if err != nil {
+			log.Fatalf("xpi.populateRsa4096Cache: %v", err)
+		}
+		s.rsa4096Cache <- key
+		time.Sleep(time.Minute)
+	}
+}
+
+// retrieve a key from the cache or generate one if it takes too long
+// or if the size is wrong
+func (s *PKCS7Signer) getRsaKey(size int) (*rsa.PrivateKey, error) {
+	switch size {
+	case 2048:
+		select {
+		case key := <-s.rsa2048Cache:
+			return key, nil
+		case <-time.After(100 * time.Millisecond):
+			// generate a key if none available
+			return rsa.GenerateKey(rand.Reader, size)
+		}
+	case 4096:
```

Comment:
nit: perhaps cleaner to keep a `map[int]chan *rsa.PrivateKey` to look up the caches instead of duplicated code

probably overkill for two key sizes, but maybe more sizes will be added in the future?

Reply:
agreed, but indeed something I'd do in a refactoring if we end up needed more key sizes. I think we'll move to ecdsa before that though.